File: viewer/src/main/scala/farjs/viewer/ViewItemsPopup.scala
package farjs.viewer

import farjs.filelist.FileListActions.FileListDirUpdatedAction
import farjs.filelist._
import farjs.filelist.api.{FileListDir, FileListItem}
import farjs.ui.Dispatch
import farjs.ui.popup._
import farjs.ui.task.{Task, TaskAction}
import farjs.viewer.ViewItemsPopup._
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.scalajs.js
import scala.util.{Failure, Success}

object ViewItemsPopup {

  private[viewer] var statusPopupComp: ReactClass = StatusPopup
}

class ViewItemsPopup(data: FileListData) extends FunctionComponent[FileListPluginUiProps] {

  protected def render(compProps: Props): ReactElement = {
    val (currDir, setCurrDir) = useState("")
    val inProgress = useRef(true)
    val props = compProps.plain

    def viewItems(dispatch: Dispatch, actions: FileListActions, state: FileListState): Unit = {
      val parent = state.currDir.path
      val currSelected = FileListState.selectedItems(state).toList
      val currItems =
        if (currSelected.nonEmpty) currSelected
        else FileListState.currentItem(state).toList
      
      var sizes = currItems.map {
        case i if i.isDir => (i.name, 0d)
        case i => (i.name, i.size)
      }.toMap
      
      val resultF = currItems.foldLeft(Future.successful(true)) { (resF, currItem) =>
        resF.flatMap {
          case true if currItem.isDir =>
            setCurrDir(currItem.name)
            var s = 0d
            actions.scanDirs(parent, js.Array(currItem), onNextDir = { (_, items) =>
              s += items.foldLeft(0d) { case (res, i) =>
                res + (if (i.isDir) 0d else i.size)
              }
              inProgress.current
            }).toFuture.map { res =>
              sizes = sizes.updated(currItem.name, s + sizes.getOrElse(currItem.name, 0d))
              res
            }
          case res => Future.successful(res)
        }
      }
      resultF.onComplete { res =>
        props.onClose()
        res match {
          case Success(false) => // already cancelled
          case Success(true) =>
            val updatedItems = state.currDir.items.map { item =>
              sizes.get(item.name) match {
                case Some(size) => FileListItem.copy(item)(size = size)
                case None => item
              }
            }
            dispatch(FileListDirUpdatedAction(FileListDir.copy(state.currDir)(items = updatedItems)))
          case Failure(_) =>
            dispatch(TaskAction(Task("Viewing Items", resultF)))
        }
      }
    }

    useLayoutEffect({ () =>
      // start scan
      viewItems(data.dispatch, data.actions, data.state)
      ()
    }, Nil)
    
    <(statusPopupComp)(^.plain := StatusPopupProps(
      text = s"Scanning the folder\n$currDir",
      title = "View",
      onClose = { () =>
        // stop scan
        inProgress.current = false
      }: js.Function0[Unit]
    ))()
  }
}

File: viewer/ViewItemsPopup.mjs
/**
 * @typedef {import("@farjs/filelist/FileListState.mjs").FileListState} FileListState
 * @import { FileListDirUpdatedAction } from "@farjs/filelist/FileListActions.mjs"
 * @import { Dispatch, FileListData } from "@farjs/filelist/FileListData.mjs"
 * @import { FileListPluginUiProps } from "@farjs/filelist/FileListPlugin.mjs"
 */
import React, { useLayoutEffect, useRef, useState } from "react";
import Task from "@farjs/ui/task/Task.mjs";
import TaskAction from "@farjs/ui/task/TaskAction.mjs";
import FileListActions from "@farjs/filelist/FileListActions.mjs";
import FileListState from "@farjs/filelist/FileListState.mjs";
import StatusPopup from "@farjs/ui/popup/StatusPopup.mjs";

const h = React.createElement;

/**
 * @param {FileListData} data
 */
function ViewItemsPopup(data) {
  /**
   * @param {FileListPluginUiProps} props
   */
  const ViewItemsPopupComp = (props) => {
    const { statusPopupComp } = ViewItemsPopup;

    const [currDir, setCurrDir] = useState("");
    const inProgress = useRef(true);

    /**
     * @param {Dispatch} dispatch
     * @param {FileListActions} actions
     * @param {FileListState} state
     * @returns {void}
     */
    function viewItems(dispatch, actions, state) {
      const parent = state.currDir.path;
      const currSelected = FileListState.selectedItems(state);
      const currItems =
        currSelected.length > 0
          ? currSelected
          : [FileListState.currentItem(state)].filter((_) => _ !== undefined);

      const sizes = currItems.reduce((res, i) => {
        return i.isDir ? res.set(i.name, 0) : res.set(i.name, i.size);
      }, /** @type {Map<string, number>} */ (new Map()));

      const resultP = currItems.reduce((resP, currItem) => {
        return resP.then((res) => {
          if (res && currItem.isDir) {
            setCurrDir(currItem.name);
            let currSize = 0;
            return actions
              .scanDirs(parent, [currItem], (_, items) => {
                currSize += items.reduce((res, i) => {
                  return res + (i.isDir ? 0 : i.size);
                }, 0);

                return inProgress.current;
              })
              .then((res) => {
                sizes.set(currItem.name, currSize);
                return res;
              });
          }

          return res;
        });
      }, Promise.resolve(true));

      resultP.then(
        (res) => {
          props.onClose();
          if (res) {
            const updatedItems = state.currDir.items.map((item) => {
              const size = sizes.get(item.name);
              return size !== undefined ? { ...item, size: size } : item;
            });
            /** @type {FileListDirUpdatedAction} */
            const action = {
              action: "FileListDirUpdatedAction",
              currDir: { ...state.currDir, items: updatedItems },
            };
            dispatch(action);
          }
        },
        () => {
          props.onClose();
          dispatch(TaskAction(Task("Viewing Items", resultP)));
        }
      );
    }

    useLayoutEffect(() => {
      // start scan
      viewItems(data.dispatch, data.actions, data.state);
    }, []);

    return h(statusPopupComp, {
      text: `Scanning the folder\n${currDir}`,
      title: "View",
      onClose: () => {
        // stop scan
        inProgress.current = false;
      },
    });
  };

  ViewItemsPopupComp.displayName = "ViewItemsPopup";

  return ViewItemsPopupComp;
}

ViewItemsPopup.statusPopupComp = StatusPopup;

export default ViewItemsPopup;
